### DISCUSSION TITLE
Does livewire support view composer variables?

### DISCUSSION THREAD

**Discussion by bilogic** [livewire/livewire#3453] [1. Help] [answered]
```PHP
// layouts/app.blade.php
<!DOCTYPE html>
<html>
    <head>
        @livewireStyles
    </head>
    <body>
        @include('heading')
        @livewire('header')
        @livewireScripts
    </body>
</html>

// heading.blade.php
{{ $user->name }}   // this will print user name

// livewire/header.blade.php
{{ $user->name }}   // this will throw an exception

// ServiceProvider
View::composer(
    ['layouts.app'],
    ViewComposer::class
);

class ViewComposer
{
    public function compose(View $view)
    {
        $view->with('user', Auth::user());
    }
}
```

Is `livewire/header.blade.php` throwing an exception by design?
Or did I do something wrong here? Thanks!

**Comment by joshhanley** [comment]
@bilogic I'm not sure if it does, but if it did, you would still need to pass the variable into the Livewire component as it has it's own context (like Blade Components do).

So change to something like this

```blade
@livewire('header', ['user' => $user])
```

But I personally wouldn't do it that way, if you need a user property in your component, just put this in your component's mount method

```php
public function mount() {
    $this->user = Auth::user();
}
```

Hope this helps!

**Reply by bilogic** [reply]
@joshhanley 

Thanks for sharing. The benefit of making `$user` available by just specifying `View::composer(['layout.app']...)` is that child snippets (`heading.blade.php` etc) have access to its parent's variables without having to explicitly declare or pass all of them in. Otherwise it would almost be like having to pass in every class variable to a class method in order to use them.

So, I find the treatment inconsistent, let me explain:

1. `@include('heading')` and `@livewire('header')` both read similarly to me, i.e. bring in the matching `blade.php` as a snippet
2. Both `View::composer(['livewire.header'], ...)` and `View::share("user", Auth::user())` works and makes `{{ $user }}` available in `livewire/header.blade.php`
3. So, Livewire components does accept variables specified outside itself
4. I would have expected them not to work if a variable must be passed into a Livewire component
5. But since they do, perhaps we could make `@livewire` behave like Laravel's `@include`?
6. This is of course bearing mind that any variable of the same name declared in the Livewire component will override the one outside

I had some trouble locating the code that does this in Laravel/Livewire, perhaps you can point me in the right direction? Thank you.

**Reply by joshhanley** [reply]
@bilogic yeah I get the benefit of view composers.

But you are comparing apples with oranges 🙂  `@livewire` is the blade equivalent of `@component`, not `@include`.

They're both components and as such have isolated scope; and contain both a php class as well as a blade view(unlike the `@include` which is just a view partial only).

Can I suggest testing how a blade `@component` works with what you hoping to achieve?

It's also why I personally prefer the component tag syntax, like `<livewire:header />` as they are components so that to me is cleaner.

Also using the view composer directly on `livewire/header.blade.php` that will probably only make the variables accessible in the blade view, but not in the component class. It's like the equivalent of passing data into the view in the render method. 
Like:
```php
public function render() {
    return view('livewire.header', ['user' => Auth::user()]);
}
```

I'd suspect that if you tried to access that user inside the Livewire component class, that it wouldn't work.

Hopefully that all makes sense. Let me know if you want me to clarify anything, or if you find different results in your experimentation 🙂

**Reply by bilogic** [reply]
@joshhanley 

Ok, I was not aware of blade `@component`, let me read it up. Off hand, I can't think of why I would need to access the variable inside the component class, it is usually just meant for `blade.php`. 

Thanks!

**Comment by bilogic** [comment, marked as answer]
@joshhanley

```PHP
// layouts/app.blade.php
<!DOCTYPE html>
<html>
    <head>
        @livewireStyles
    </head>
    <body>
        @include('heading')
        @livewireScripts
    </body>
</html>

// heading.blade.php
{{ $user->name }}   // this will print user name
...
@livewire('cart')
...
```
So what I ended up doing was to take out the bits that could use @include, and leave only 1 tag as a @livewire(...). Works beautifully now!

I will mark this as the answer, but credit goes to @joshhanley for the inspiration after I spent close to a day on this. Thank you!
